src/annotations/components/annotationForm/AnnotationForm.test.tsx

```diff
@@ -0,0 +1,226 @@
+import React from 'react'
+import {renderWithRedux} from 'src/mockState'
+import {
+  AnnotationForm,
+  END_TIME_IN_FUTURE_MESSAGE,
+  START_TIME_IN_FUTURE_MESSAGE,
+  TIMES_ARE_SAME_MESSAGE,
+  WRONG_ORDER_MESSAGE,
+} from 'src/annotations/components/annotationForm/AnnotationForm'
+import {fireEvent} from '@testing-library/react'
+import {createDateTimeFormatter} from 'src/utils/datetime/formatters'
+import {REQUIRED_ERROR} from 'src/annotations/components/annotationForm/AnnotationTimeInput'
+
+const setup = (annoType: 'point' | 'range', startTime, endTime) => {
+  const doNothing = () => {}
+
+  return renderWithRedux(
+    <AnnotationForm
+      title="Add"
+      type={annoType}
+      startTime={startTime}
+      endTime={endTime}
+      eventPrefix="testing"
+      onClose={doNothing}
+      onSubmit={doNothing}
+    />
+  )
+}
+
+const setupPoint = time => {
+  return setup('point', time, null)
+}
+
+const setupRange = (start, end) => {
+  return setup('range', start, end)
+}
+
+describe('Annotation Form Component tests', () => {
+  const fiveMinutesAgo = new Date().getTime() - 1000 * 60 * 5
+  const oneHourAgo = new Date().getTime() - 1000 * 60 * 60
+  const oneHourInFuture = fiveMinutesAgo + 60 * 60 * 1000
+
+  const formatter = createDateTimeFormatter('YYYY-MM-DD hh:mm:ss a')
+
+  it('should not allow an annotation without a message', () => {
+    const {queryByTestId, getByTestId} = setupPoint(fiveMinutesAgo)
+
+    expect(getByTestId('form--element-error')).toHaveTextContent(
+      'This field is required'
+    )
+    expect(getByTestId('annotation-submit-button')).toBeDisabled()
+
+    fireEvent.change(getByTestId('edit-annotation-message'), {
+      target: {value: 'abcde'},
+    })
+
+    expect(queryByTestId('form--element-error')).toBeNull()
+    expect(getByTestId('annotation-submit-button')).not.toBeDisabled()
```

Comment:
so i think as a general rule, unit / integration tests should only test one action. this is testing two actions - the failure of the validation, and the success of it being fixed. what do you think of having a describe section for the validation failures and then having one for it fixing it?

i was mainly reading this thread: https://softwareengineering.stackexchange.com/a/171502/10616

Reply:
that would be a duplication of tests, since need to verify that it is invalid , and then that fixing it makes it valid.

(test a, then second test: a and b.... where the second test supersedes the first test)

Reply:
I still prefer that to having multiple actions packed into a test, but that's an ideal to strive for.

If you prefer to keep it this way, can you update the test descriptions in these to say that they also fix the validation issues?

Reply:
i'll update the descriptions.

Reply:
...updated the description at the top, since each test messes up the validation, checks that the button is disabled, then fixes it and confirms that the validation message is gone and that the button is enabled.